### PR TITLE
Foreignkey

### PR DESCRIPTION
**Pull request by LiuXiaoxuanPKU**

### CODE REVIEW

File: constropt/constr_extractor/builtin_extractor.rb [hunk 1]
```diff
@@ -242,117 +254,120 @@ def extract_builtin_numerical(ast)
...
-        k, v = handle_assoc_node(n)
-        if !k.nil? && k == 'class_name'
-          fields[0] = trim_string(v.source.to_s)
-        end
-        if !k.nil? && k == 'foreign_key'
-          foreign_key = trim_string(v.source.to_s)
-        end
-      end
+    if class_name.nil?
+      class_name = fields[0].capitalize
```
[the remark this answers is not in the document]
Comment (sharonlee2110) [reply constropt/constr_extractor/builtin_extractor.rb:331]: This is because if class_name or foreign_key is not specified, the ruby documentation I found here [https://guides.rubyonrails.org/association_basics.html](https://guides.rubyonrails.org/association_basics.html ) says that the default is based on the field name.

[the remark this answers is not in the document]
Comment (sharonlee2110) [reply constropt/constr_extractor/builtin_extractor.rb:331]: There is only one field, but looping through all the nodes in content allows me to parse each node and at the same time extract the field

File: constropt/constr_extractor/relationship_extractor.rb [hunk 2]
```diff
@@ -0,0 +1,74 @@
+#copy over the relationship data in the initialization function
```
[the remark this answers is not in the document]
Comment (sharonlee2110) [reply constropt/constr_extractor/relationship_extractor.rb:1]: For now this file is not being used

File: constropt/constr_extractor/test/test_engine.rb [hunk 3]
```diff
@@ -13,7 +14,9 @@ def visit(node, _params)
     inclusion_constraints = node.constraints.select { |c| c.is_a? InclusionConstraint }
     unique_constraints = node.constraints.select { |c| c.is_a? UniqueConstraint }
     foreign_key_constraints = node.constraints.select { |c| c.is_a? ForeignKeyConstraint }
-    (inclusion_constraints + unique_constraints + foreign_key_constraints).each do |c|
+    presence_constraints = node.constraints.select { |c| c.is_a? PresenceConstraint }
```
[the remark this answers is not in the document]
Comment (sharonlee2110) [reply constropt/constr_extractor/test/test_engine.rb]: I wanted to cross-check the number of foreign key constraints extracted from redmine with the expected number

File: constropt/constr_extractor/builtin_extractor.rb [hunk 4]
```diff
@@ -2,9 +2,14 @@
...
 require_relative 'constraint'
 
+def trim_string(s)
+  s.delete_prefix("'").delete_suffix("'").delete_prefix('"').delete_suffix('"')
+end
+
 class BuiltinExtractor < Extractor
   BUILTIN_VALIDATOR = %w[validates_presence_of validates_uniqueness_of
                          validates_format_of validates_length_of
+                         validates_inclusion_of belongs_to
```
Comment (LiuXiaoxuanPKU) [inline constropt/constr_extractor/builtin_extractor.rb:12]: two validates_inclusion_of here

File: constropt/constr_extractor/builtin_extractor.rb [hunk 5]
```diff
@@ -274,7 +284,7 @@ def extract_numerical_hash(node)
         max = v[0].to_i if !k.nil? && %w[less_than_or_equal_to less_than].include?(k)
       rescue StandardError
         puts "[Error] Fail to parse min, max value k: #{k}, v: #{v.source}"
-      end
+        end
```
Comment (LiuXiaoxuanPKU) [inline constropt/constr_extractor/builtin_extractor.rb:287]: wrong indent here

File: constropt/constr_extractor/builtin_extractor.rb [hunk 6]
```diff
@@ -57,6 +62,11 @@ def extract_bultin_validator(ast)
       constraints = extract_builtin_length(ast)
     when 'validates_numericality_of'
       constraints = extract_builtin_numerical(ast)
+    when 'belongs_to'
+      # extract_builtin_belongs_to(ast)
+      constraints = extract_builtin_foreign(ast)
+    when 'state_machine'
```
Comment (LiuXiaoxuanPKU) [inline constropt/constr_extractor/builtin_extractor.rb:68]: state machine is in another pass, so just remove state_machine here

File: constropt/constr_extractor/relationship_extractor.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (LiuXiaoxuanPKU) [reply constropt/constr_extractor/relationship_extractor.rb:1]: If it is not used, could you change filename to relationship_extractor_unused.rb

File: extract.rb [hunk 7]
```diff
@@ -6,13 +6,15 @@
 require_relative 'constropt/constr_extractor/class_inheritance_extractor'
 require_relative 'constropt/constr_extractor/constraint'
 require_relative 'constropt/constr_extractor/serializer'
+#require_relative 'constropt/constr_extractor/relationship_extractor'
```
Comment (LiuXiaoxuanPKU) [inline extract.rb:9]: Could you revert the change of this file?
